### ISSUE TITLE
[Q] wandb: Error('You must call wandb.init() before wandb.log()')"

### ISSUE THREAD

**Issue by Inkalika** [wandb/wandb#6540]
I want to choose hyperparameters for my metric (so far I have chosen the metric of the model itself). I use wandb sweep for this purpose. I tried the code. which is given below, but I get an error **_"wandb: Error('You must call wandb.init() before wandb.log()')"_**
```
import wandb
import ultralytics
from ultralytics import YOLO

sweep_configuration = {
    "method": "random",
    "name": "sweep",
    "metric": {"goal": "maximize", "name": "acc_top1"},
    'parameters': {
        'batch': {'values': [16, 32]},  
        'imgsz': {'values': [224, 256]}
        'epochs': {'distribution': 'int_uniform', 'min': 1, 'max': 3} 
    }
}

sweep_id = wandb.sweep(sweep=sweep_configuration, project="my-first-sweep")

def main():
    wandb.init()
        config = wandb.config
        model = YOLO('yolov8n-cls.pt', task='classify')
        results = model.train(data='C:/Users/Downloads/', **config, plots=False, save=False, val=True)   
        acc_top1 = results.results_dict['metrics/accuracy_top1']
        wandb.log({"acc_top1": acc_top1})

wandb.agent(sweep_id, function=main, count=2)
```

I tried another code, but I get the following error **_"wandb: UsageError('Run (yfaxga7o) is finished. The call to `log` will be ignored. Please make sure that you are using an active run.')"_**
```
import wandb
import ultralytics
from ultralytics import YOLO

sweep_configuration = {
    "method": "random",
    "name": "sweep",
    "metric": {"goal": "maximize", "name": "acc_top1"},
    'parameters': {
        'batch': {'values': [16, 32]},  
        'imgsz': {'values': [224, 256]}
        'epochs': {'distribution': 'int_uniform', 'min': 1, 'max': 3}         
    }
}

sweep_id = wandb.sweep(sweep=sweep_configuration, project="my-first-sweep")

def main():
    with wandb.init() as run:
        config = run.config
        model = YOLO('yolov8n-cls.pt', task='classify')
        results = model.train(data='C:/Users/Downloads/', **config, plots=False, save=False, val=True)   
        acc_top1 = results.results_dict['metrics/accuracy_top1']
        run.log({"acc_top1": acc_top1})

wandb.agent(sweep_id, function=main, count=2)
```

Because of these errors, optimization is not performed and my metric is not sent to wandb.ai

How can you still choose hyperparameters to optimize your metric, except using the _"tune"_ method?

**Comment by anmolmann** [comment]
Hey @Inkalika , within your Python script, you can [define a sweep config](https://docs.wandb.ai/guides/sweeps/define-sweep-configuration) in a dictionary Python data structure or pass them in a YAML file. Feel free to add your additional 
 parameter key-value pairs to the `parameters` key in the dictionary.

Also, the code you shared above gave some errors, by resolving them - I was able to start a sweep and launch sweep agents on my end. Please update your `PROJECT_NAME`, `ENTITY_NAME` and `PATH_TO_DATA` in the snippet below and execute the script.
```
import wandb
import ultralytics
from ultralytics import YOLO

sweep_configuration = {
    "method": "random",
    "name": "sweep",
    "metric": {"goal": "maximize", "name": "acc_top1"},
    'parameters': {
        'batch': {'values': [16, 32]},  
        'imgsz': {'values': [224, 256]},
        'epochs': {'distribution': 'int_uniform', 'min': 1, 'max': 3} 
    }
}

sweep_id = wandb.sweep(sweep=sweep_configuration, project="<PROJECT_NAME>", entity="<ENTITY_NAME>")

def main():
    wandb.init()
    config = wandb.config
    model = YOLO('yolov8n-cls.pt', task='classify')
    results = model.train(data='<PATH_TO_DATA>', **config, plots=False, save=False, val=True)
    acc_top1 = results.results_dict['metrics/accuracy_top1']
    wandb.log({"acc_top1": acc_top1})

wandb.agent(sweep_id, function=main, count=2)
```

Also, to pass in another set of config, you can also check out our `command` [configuration for sweeps](https://docs.wandb.ai/guides/sweeps/define-sweep-configuration#command-). To add extra command line arguments not specified by sweep configuration parameters:
```
command:
  - ${env}
  - ${interpreter}
  - ${program}
  - "--config"
  - "your-training-config.json"
  - ${args}
```

**Comment by Inkalika** [comment]
![image](https://github.com/wandb/wandb/assets/89869109/b454e6fc-7663-407d-924d-2713b33c9183)
Hello, @anmolmann 
The error I have is that I can't get acc_top1 graphs. I think this error is related to the work of YOLO. Apparently, after completing its part of the work, it closes the connection with wandb.init() and the log is not generated and I don’t know how to fix it.
`adding sweep_id = wandb.sweep(sweep=sweep_configuration, project="<PROJECT_NAME>", entity="<ENTITY_NAME>")` to the code creates the error ```
"404 response executing GraphQL.
{"errors":[{"message":"entity not found during upsertSweep","path":["upsertSweep"]}],"data":{"upsertSweep":null}}
wandb: ERROR Error while calling W&B API: entity not found during upsertSweep (<Response [404]>)"
```

**Comment by anmolmann** [comment]
Hey @Inkalika, the above error: `entity not found during upsertSweep` seems to indicate that the entity is not configured correctly. Have you set the env var `WANDB_ENTITY` via the env var on your end? You can do this by running this export cmd:
`export WANDB_ENTITY=<your_entity>`, verify this by `echo $WANDB_ENTITY`.

**Comment by anmolmann** [comment]
Hi @Inkalika , I wanted to follow up on this request. Please let us know if we can be of further assistance or if your issue has been resolved.

**Comment by sydholl** [comment]
WandB Internal User commented: 
anmolmann commented: 
Hi @Inkalika, I wanted to follow up on this request. Please let us know if we can be of further assistance or if [your issue ](https://github.com/wandb/wandb/issues/6540#issuecomment-1819795707)has been resolved once you set the project and the entity via env vars.

**Comment by sydholl** [comment]
WandB Internal User commented: 
Inkalika commented: 
I couldn't fix the bug, that's why I used the Optuna framework. Thank you

**Comment by sydholl** [comment]
WandB Internal User commented: 
sydholl commented: 
WandB Internal User commented: 
anmolmann commented: 
Hi @Inkalika, I wanted to follow up on this request. Please let us know if we can be of further assistance or if [your issue ](https://github.com/wandb/wandb/issues/6540#issuecomment-1819795707)has been resolved once you set the project and the entity via env vars.

**Comment by sydholl** [comment]
WandB Internal User commented: 
Inkalika commented: 
I couldn't fix the bug, that's why I used the Optuna framework. Thank you

**Comment by sydholl** [comment]
WandB Internal User commented: 
sydholl commented: 
WandB Internal User commented: 
Inkalika commented: 
I couldn't fix the bug, that's why I used the Optuna framework. Thank you

**Comment by sydholl** [comment]
WandB Internal User commented: 
sydholl commented: 
WandB Internal User commented: 
Inkalika commented: 
I couldn't fix the bug, that's why I used the Optuna framework. Thank you

**Comment by anmolmann** [comment]
@Inkalika, I see. Apologies for the inconvenience caused due to this. Would you like to schedule a call to debug this issue live instead just in case?

**Comment by Inkalika** [comment]
Thank you. I think there is no need. I think this problem is related to YOLO8 itself. It also calls the wandb function when executing its code. Perhaps because of this there is a conflict. But if this code still works on YOLO8, and does not work only on my computer, then it means only starting the installation from the beginning and trying again

**Comment by anmolmann** [comment]
I see, thanks for the update @Inkalika ! Please let me know if you try reinstalling from the beginning again.
Also, we have a [report detailing](https://api.wandb.ai/links/geekyrakshit/01j7zv38) how to integrate yolov8 with wandb as well for additional context and reference.

ultralytics==8.0.186
https://github.com/wandb/wandb/issues/6412#issuecomment-1746850406

[WandB integration is available out of the box in YOLOv8](https://github.com/ultralytics/ultralytics/issues/848#issuecomment-1777857811), so there's no need for an additional setup. To use it, you just need the wandb library installed in your working environment. When it's present, YOLOv8 will automatically detect it and log your training runs to your WandB account.

This automatic logging includes metrics like loss and accuracy, along with other training and validation statistics. It's a powerful tool for tracking your model's performance over time, experimenting with hyperparameters, and comparing different models.

Ensure you're logged into WandB through your command line before starting your training for your runs to be tracked. Do note that different versions could influence the way YOLOv8 interacts with WandB, so make sure to keep your packages updated.

I hope this information helps! If you encounter any issue, please feel free to provide specific details so we can give you more targeted guidance. Happy training!

Please let us know if you would like to reopen the issue.
